storm-client/test/jvm/org/apache/storm/topology/PersistentWindowedBoltExecutorTest.java

```diff
@@ -0,0 +1,261 @@
+
+        Mockito.doAnswer(new Answer<Void>() {
+            @Override
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+                TupleWindow window = (TupleWindow) invocation.getArguments()[0];
+                // iterate the tuples
+                Assert.assertEquals(5, window.get().size());
+                // iterating multiple times should produce same events
+                Assert.assertEquals(5, window.get().size());
+                Assert.assertEquals(5, window.get().size());
+                return null;
+            }
+        }).when(mockBolt).execute(Mockito.any());
+        // trigger the window
+        long activationTs = tupleTs + 1000;
+        executor.getWindowManager().add(new WaterMarkEvent<>(activationTs));
+        executor.prePrepare(0);
+
+        // partition ids
+        Mockito.verify(mockPartitionState, Mockito.times(1)).put(stringCaptor.capture(), partitionValuesCaptor.capture());
+        Assert.assertEquals(PARTITION_KEY, stringCaptor.getAllValues().get(0));
+        Assert.assertEquals(Collections.singletonList(0L), partitionValuesCaptor.getAllValues().get(0));
+
+        // window partitions
+        Mockito.verify(mockWindowState, Mockito.times(1)).put(longCaptor.capture(), windowValuesCaptor.capture());
+        Assert.assertEquals(0L, (long) longCaptor.getAllValues().get(0));
+        Assert.assertEquals(5, windowValuesCaptor.getAllValues().get(0).size());
+        List<Event<?>> tuples = windowValuesCaptor.getAllValues().get(0)
+            .getEvents().stream().map(Event::get).collect(Collectors.toList());
+        Assert.assertArrayEquals(mockTuples.toArray(), tuples.toArray());
+
+        // window system state
+        Mockito.verify(mockSystemState, Mockito.times(2)).put(stringCaptor.capture(), systemValuesCaptor.capture());
+        Assert.assertEquals(EVICTION_STATE_KEY, stringCaptor.getAllValues().get(1));
+        Assert.assertEquals(Optional.of(Pair.of(5L, 5L)), systemValuesCaptor.getAllValues().get(0));
+        Assert.assertEquals(TRIGGER_STATE_KEY, stringCaptor.getAllValues().get(2));
+        Assert.assertEquals(Optional.of(tupleTs), systemValuesCaptor.getAllValues().get(1));
+    }
+
+    @Test
+    public void testCacheEviction() {
+        Mockito.when(mockWaterMarkEventGenerator.track(Mockito.any(GlobalStreamId.class), Mockito.anyLong())).thenReturn(true);
+        executor.initState(null);
+        executor.waterMarkEventGenerator = mockWaterMarkEventGenerator;
+        List<Tuple> mockTuples = getMockTuples(20000);
+        mockTuples.forEach(t -> executor.execute(t));
+
+        Mockito.verify(mockWindowState, Mockito.times(10)).put(longCaptor.capture(), windowValuesCaptor.capture());
+        Assert.assertEquals(10, longCaptor.getAllValues().size());
+        Assert.assertEquals(10, windowValuesCaptor.getAllValues().size());
+        // number of evicted events
+        Assert.assertEquals(10_000, windowValuesCaptor.getAllValues().stream()
```

Comment:
Here its checking if the `PersistentWindowedBoltExecutor` is writing the evicted values out to the state. Will add separate unit tests for WindowState.